### PR TITLE
Coreclr compatibility: ApplicationException -> Exception

### PR DESCRIPTION
**Pull request by dVakulen**
Related to #368

### CODE REVIEW

File: src/OrleansAzureUtils/Storage/AzureBasedReminderTable.cs [hunk 1]
```diff
@@ -89,7 +89,7 @@ private ReminderEntry ConvertFromTableEntry(ReminderTableEntry tableEntry, strin
                     var error = String.Format( "Read a reminder entry for wrong Service id. Read {0}, but my service id is {1}. Going to discard it.",
                         tableEntry, serviceIdStr);
                     logger.Warn(ErrorCode.AzureTable_ReadWrongReminder, error);
-                    throw new ApplicationException(error);
+                    throw new Exception(error);
```
Comment (gabikliot) [inline src/OrleansAzureUtils/Storage/AzureBasedReminderTable.cs]: can you lease use `OrleansException` here instead?

File: src/ClientGenerator/NamespaceGenerator.cs [hunk 2]
```diff
@@ -392,7 +392,7 @@ protected string GetActualMethodReturnType(Type type, SerializeFlag flag)
 
                 var errorMsg = string.Format("Unexpected number of arguments {0} for generic type {1} used as a return type. Only Type<T> are supported as generic return types of grain methods.", genericArguments.Length, type);
                 ReportError(errorMsg);
-                throw new ApplicationException(errorMsg);
+                throw new Exception(errorMsg);
```
Comment (gabikliot) [inline src/ClientGenerator/NamespaceGenerator.cs]: Can you please use ArgumentException here instead?

**Comment by gabikliot** [comment]
@dVakulen , thank you very much with your help on #368!

We should probably not do any work on `Type` conversion at this point, since hopefully PR #528 (lead by @ReubenBond) will deal with that natively and be coreclr complaint. Even if not, the code gen (which is the majority of the code that uses those non coreclr APIs) will change a lot anyway with #528 .
Once #528 is accepted, which is hopefully pretty soon, we should re-run the API portability tool and see what is remaining.

File: src/ClientGenerator/NamespaceGenerator.cs (hunk 2 above)
Comment (ReubenBond) [inline src/ClientGenerator/NamespaceGenerator.cs]: This file will be removed in a PR which will shortly follow #528

**Comment by dVakulen** [comment]
Ok, then I'll continue work on сoreclr compatibility after merge of PR #528.

**Comment by gabikliot** [comment]
I was only referring to `Type.IsInterface-> Type.GetTypeInfo.IsInterface` (and similar issues) and mostly all the code in the ClientGenerator solution and the Serialization. Only this code will be impacted by #528. All the other coreclr compatibility issues will not be fixed by #528.

**Comment by gabikliot** [comment]
Looks good to merge.

File: src/ClientGenerator/NamespaceGenerator.cs [hunk 3]
```diff
@@ -392,7 +392,7 @@ protected string GetActualMethodReturnType(Type type, SerializeFlag flag)
 
                 var errorMsg = string.Format("Unexpected number of arguments {0} for generic type {1} used as a return type. Only Type<T> are supported as generic return types of grain methods.", genericArguments.Length, type);
                 ReportError(errorMsg);
-                throw new ApplicationException(errorMsg);
+                throw new ArgumentException(errorMsg);
```
Comment (sergeybykov) [inline src/ClientGenerator/NamespaceGenerator.cs]: Another minor thing - per .NET design guidelines it should be `throw new ArgumentException(errorMsg, "type");` to indicate which argument is invalid.

Comment (dVakulen) [inline src/ClientGenerator/NamespaceGenerator.cs]: Fixed

**Comment by sergeybykov** [comment]
Thank you, @dVakulen!

**Pull request merged**